universal/include/userver/formats/universal/common_containers.hpp

```diff
@@ -0,0 +1,296 @@
+    }
+    return std::nullopt;
+  }
+  constexpr auto Write(const std::optional<std::string>& value, std::string_view fieldName, const auto&, auto& builder) const {
+    if(value) {
+      builder[static_cast<std::string>(fieldName)] = *value;
+    };
+  };
+  inline constexpr std::string_view Check(const std::string&) const {
+    return "";
+  }
+
+};
+template <>
+struct FieldConfig<std::string> {
+  std::optional<Pattern> Pattern;
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::string Read(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    return value[name].template As<std::string>();
+  };
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<std::string> TryRead(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    return parse::TryParse(value[name], parse::To<std::string>{});
+  };
+  constexpr auto Write(std::string_view value, std::string_view fieldName, const auto&, auto& builder) const {
+    builder[static_cast<std::string>(fieldName)] = value;
+  };
+  inline constexpr std::string_view Check(std::string_view) const {
+    return "";
+  }
+
+};
+
+template <>
+struct FieldConfig<std::optional<int>> {
+  std::optional<Max<int>> Maximum;
+  std::optional<Min<int>> Minimum;
+  std::optional<Default<int>> Default;
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<int> Read(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    if(value.HasMember(name)) {
+      return value[name].template As<int>();
+    }
+    if(this->Default) {
+      return this->Default->value;
+    }
+    return std::nullopt;
+  }
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<std::optional<int>> TryRead(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    auto response = parse::TryParse(value[name], parse::To<int>{});
+    if(response) {
+      return response;
+    }
+    if(this->Default) {
+      return this->Default->value;
+    }
+    return {{}};
+  }
+  constexpr auto Write(const std::optional<int>& value, std::string_view fieldName, const auto&, auto& builder) const {
+    if(value) {
+      builder[static_cast<std::string>(fieldName)] = *value;
+      return;
+    }
+    if(this->Default) {
+      builder[static_cast<std::string>(fieldName)] = this->Default->value;
+    }
+  }
+
+  inline constexpr std::string_view Check(const std::optional<int>&) const {
+    return "";
+  }
+
+};
+template <typename Value>
+struct FieldConfig<std::unordered_map<std::string, Value>> {
+  std::optional<Additional> Additional;
+  using kType = std::unordered_map<std::string, Value>;
+  template <typename MainClass, auto I, typename Value2>
+  inline constexpr kType Read(Value2&& value) const {
+    if(!this->Additional) {
+      throw std::runtime_error("Invalid Flags");
```

Comment:
A meaning can be given to map that is `!Additional`: it's parsed from a single field like a normal nested map, e.g.

```cpp
struct SomeStruct {
  int foo;
  std::unordered_map<std::string, int> extra;
};
```

With `Additional{true}`, it has this serialized format:

```json
{
  "foo": 42,
  "bar": 10,
  "baz": 20
}
```

With `Additional{false}`, it has this serialized format:

```json
{
  "foo": 42,
  "extra": {
    "bar": 10,
    "baz": 20
  }
}
```